workflow/workflow.go

```diff
@@ -325,31 +349,91 @@ func (l *loopState) onStageComplete(stepID string, previousStage *string, previo
 	l.notifySteps()
 }
 
+// Marks the outputs of that stage unresolvable.
+// Optionally skip an output that should not unresolve.
+// Does not (un)resolve the non-output node for that stage.
+// To prevent a deadlock, this `notifySteps()` should be called at some point after this is called.
+func (l *loopState) unresolveOutputs(stepID string, stageID string, skippedOutput *string) {
+	stages := l.lifecycles[stepID].Stages
+	for _, stage := range stages {
+		if stage.ID != stageID {
+			continue
+		}
+		for stageOutputID := range stage.Outputs {
+			if skippedOutput != nil && stageOutputID == *skippedOutput {
+				continue
+			}
+			unresolvableOutputNode, err := l.dag.GetNodeByID(GetOutputNodeID(stepID, stage.ID, stageOutputID))
+			if err != nil {
+				l.logger.Warningf("Could not get DAG node %s (%s)", stepID+"."+stage.ID+"."+stageOutputID, err.Error())
+				continue
+			}
+			l.logger.Debugf("Will unresolve node %s in the DAG", stepID+"."+stage.ID+"."+stageOutputID)
+			err = unresolvableOutputNode.ResolveNode(dgraph.Unresolvable)
+			if err != nil {
+				l.logger.Errorf("Error while unresolving node %s in DAG (%s)", unresolvableOutputNode.ID(), err.Error())
+			}
+		}
+	}
+}
+
+// unresolves the stage node for the step's stage.
+// Does not (un)resolve the outputs of that node. For that, call unresolveOutputs() instead or
+// in addition to calling this function.
+// To prevent a deadlock, this `notifySteps()` should be called at some point after this is called.
+func (l *loopState) unresolveStageNode(stepID string, stageID string) {
```

Comment:
similar question, what does the verb `unresolve` mean in the arcaflow domain?

Reply:
I updated it to have a less ambiguous name.